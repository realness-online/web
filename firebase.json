{
  "storage": {
    "rules": "storage.rules"
  },
  "hosting": {
    "public": "dist",
    "ignore": [
      "firebase.json",
      "**/.*",
      "**/node_modules/**"
    ],
    "rewrites": [
      {
        "source": "**",
        "destination": "/index.html"
      }
    ],
    "headers": [
      {
        "source": "**",
        "headers": [
          {
            "key": "Content-Security-Policy",
            "value": "script-src 'unsafe-eval' 'unsafe-inline' 'self' https://www.google.com https://apis.google.com https://www.gstatic.com https://storage.googleapis.com/"
          }
        ]
      },
      {
        "source": "**",
        "headers": [
          {
            "key": "Cache-Control",
            "value": "max-age=31536000"
          }
        ]
      },
      {
        "source": "/service.worker.js",
        "headers": [
          {
            "key": "Cache-Control",
            "value": "no-cache"
          }
        ]
      }
    ]
  }
}
