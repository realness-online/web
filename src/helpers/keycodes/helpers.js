
const FORBIDDEN_NODES = ['INPUT', 'TEXTAREA', 'SELECT']

const are_objects_equal = (a, b) =>
  Object.entries(a).every(([key, value]) => b[key] === value)

export const split_combination = combination => {
  combination = combination.replace(/\s/g, '')
  combination = combination.includes('numpad+')
    ? combination.replace('numpad+', 'numpadadd')
    : combination
  combination = combination.includes('++')
    ? combination.replace('++', '+=')
    : combination
  return combination.split(/\+{1}/)
}

export const return_char_code = key => key.length === 1 ? key.charCodeAt(0) : undefined

const get_hotkey_callback = (key_map, key_code, event_key_modifiers) => {
  const key = key_map.find(({ code, modifiers }) =>
    code === key_code && are_objects_equal(event_key_modifiers, modifiers)
  )
  if (!key) return false
  return key.callback
}

export const assign_key_handler = (e, key_map, modifiers) => {
  const { keyCode: key_code, ctrlKey, altKey, shiftKey, metaKey } = e
  const event_key_modifiers = { ctrlKey, altKey, shiftKey, metaKey }

  if (modifiers.prevent) {
    e.preventDefault()
  }

  if (modifiers.stop) {
    e.stopPropagation()
  }

  const { nodeName: name, isContentEditable: editable } = document.activeElement
  if (editable || FORBIDDEN_NODES.includes(name)) return

  const callback = get_hotkey_callback(keyMap, key_code, event_key_modifiers)
  if (!callback) return e
  e.preventDefault()
  callback[e.type](e)
}
