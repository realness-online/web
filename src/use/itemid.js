// {base_url}/{:author}/{:type}/{:created_at}
import get_item from '@/use/item'
import { does_not_exist } from '@/use/sync'
import { current_user, url, directory } from '@/use/serverless'
import { get, set, keys } from 'idb-keyval'
class Directory {
  constructor() {
    this.items = []
    this.types = [] // folders are types in our vocabulary
  }
}
// Expensive to call
export async function load(itemid, me = localStorage.me) {
  let item
  if (~itemid.indexOf(me)) {
    item = localStorage.getItem(itemid)
    if (item) return get_item(item)
    else if (as_type(itemid) === 'relations') return []
  }
  const result = await get(itemid)
  item = get_item(result)
  if (item) return item
  item = await load_from_network(itemid, me)
  if (item) return item
  return null
}
export async function list(itemid, me = localStorage.me) {
  try {
    const item = await load(itemid, me)
    if (item) return type_as_list(item)
    else return []
  } catch {
    return []
  }
}
export async function load_from_network(itemid, me = localStorage.me) {
  const url = await as_download_url(itemid, me)
  if (url) {
    console.info('download', itemid)
    const server_text = await (await fetch(url)).text()
    await set(itemid, server_text)
    return get_item(server_text)
  } else return null
}
export async function load_directory_from_network(itemid) {
  if (navigator.onLine && current_user.value) {
    const path = as_directory_id(itemid)
    const meta = new Directory()
    console.info('request:directory', itemid)
    const folder = await directory(`people/${path}`)
    folder.items.forEach(item => meta.items.push(item.name.split('.')[0]))
    folder.prefixes.forEach(prefix => meta.types.push(prefix.name))
    await set(path, meta)
    return meta
  }
}
export async function as_directory(itemid) {
  const path = as_directory_id(itemid)
  const cached = await get(path)
  if (cached) return cached
  let directory = await build_local_directory(itemid)
  if (navigator.onLine && current_user.value)
    directory = await load_directory_from_network(itemid)
  return directory
}
export async function as_download_url(itemid) {
  if (!current_user.value) return null
  try {
    console.info('request:location', itemid)
    return await url(as_filename(itemid))
  } catch (e) {
    if (e.code === 'storage/object-not-found') {
      console.warn(itemid, '=>', as_filename(itemid))
      const index = (await get('sync:index')) || {}
      index[itemid] = does_not_exist
      await set('sync:index', index)
      return null
    } else throw e
  }
}

// Cheap to call
const created_at = ['posters', 'animations']
const has_history = ['statements', 'events']
export function is_history(itemid) {
  const parts = as_path_parts(itemid)
  if (has_history.includes(as_type(itemid)) && parts.length === 3) return true
  else return false
}
export function as_filename(itemid) {
  let filename = itemid
  if (itemid.startsWith('/+')) filename = `/people${filename}`
  if (created_at.includes(as_type(itemid)) || is_history(itemid))
    return `${filename}.html`
  else return `${filename}/index.html`
}
export function as_storage_path(itemid) {
  const path_parts = as_path_parts(itemid)
  let path = `/${path_parts[0]}`
  if (itemid.startsWith('/+')) path = `/people${path}`
  switch (path_parts.length) {
    case 0:
      return null
    case 1:
      return path
    default:
      return `${path}/${path_parts[1]}`
  }
}
export async function build_local_directory(itemid) {
  const path = as_directory_id(itemid)
  const directory = new Directory()
  const everything = await keys()
  everything.forEach(itemid => {
    if (as_directory_id(itemid) === path) {
      const id = as_created_at(itemid)
      if (id) directory.items.push(id)
    }
  })
  return directory
}
export function as_directory_id(itemid) {
  const parts = as_path_parts(itemid)
  return `/${parts[0]}/${parts[1]}/`
}
export function as_path_parts(itemid) {
  const path = itemid.split('/')
  if (path[0].length === 0) path.shift()
  return path
}
export function as_author(itemid) {
  const path = as_path_parts(itemid)
  const author = path[0] || ''
  if (author.startsWith('+1')) return `/${path[0]}`
  else return null
}
export function as_type(itemid) {
  const path = as_path_parts(itemid)
  if (path[1]) return path[1]
  if (itemid.startsWith('/+')) return 'person'
  else return null
}
export function as_created_at(itemid) {
  const path = as_path_parts(itemid)
  if (path.length === 1) return parseInt(path[0])
  return parseInt(path[2])
}
export function as_query_id(itemid) {
  return itemid.substring(2).replace('/', '-').replace('/', '-')
}
export function as_fragment_id(itemid) {
  return `#${as_query_id(itemid)}`
}
export function type_as_list(item) {
  // Returns a list even if loading the item fails
  // the microdata spec requires properties values to
  // single value and iterable
  if (!item) return []
  const list = item[as_type(item.id)]
  if (list && Array.isArray(list)) return list
  else if (list) return [list]
  else return []
}
