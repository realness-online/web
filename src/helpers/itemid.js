// {base_url}/{:author}/{:type}/{:created_at}
import firebase from 'firebase/app'
import 'firebase/storage'
import 'firebase/auth'
import get_item from '@/modules/item'
import { does_not_exist } from '@/persistance/Cloud.sync'
import { get, set } from 'idb-keyval'
// Expensive to call
export async function load(itemid, me = localStorage.me) {
  let item
  if (~itemid.indexOf(me)) {
    item = localStorage.getItem(itemid)
    if (item) return get_item(item)
    else if (as_type(itemid) === 'relations') return []
  }
  const result = await get(itemid)
  item = get_item(result)
  if (item) return item
  item = await load_from_network(itemid, me)
  if (item) return item
  return null
}
export async function list(itemid, me = localStorage.me) {
  try {
    const item = await load(itemid, me)
    if (item) return type_as_list(item)
    else return []
  } catch {
    return []
  }
}
export async function load_from_network(itemid, me = localStorage.me) {
  const url = await as_download_url(itemid, me)
  if (url) {
    console.info('download', itemid)
    const server_text = await (await fetch(url)).text()
    await set(itemid, server_text)
    return get_item(server_text)
  } else return null
}
export async function as_directory(itemid) {
  const path = as_directory_id(itemid)
  const cached = await get(path)
  if (cached) return cached
  else if (navigator.onLine && firebase.auth().currentUser) {
    const meta = { items: [], types: [] } // folders are types in our vocabulary
    console.info('request:directory', itemid)
    const directory = await firebase.storage().ref().child(`people/${path}`).listAll()
    directory.items.forEach(item => meta.items.push(item.name.split('.')[0]))
    directory.prefixes.forEach(prefix => meta.types.push(prefix.name))
    await set(path, meta)
    return meta
  } else return null
}
export async function as_download_url(itemid) {
  if (!firebase.auth().currentUser) return null
  const storage = firebase.storage().ref()
  try {
    console.info('request:location', itemid)
    return await storage.child(as_filename(itemid)).getDownloadURL()
  } catch (e) {
    if (e.code === 'storage/object-not-found') {
      console.warn(itemid, '=>', as_filename(itemid))
      const index = (await get('sync:index')) || {}
      index[itemid] = does_not_exist
      await set('sync:index', index)
      return null
    } else throw e
  }
}

// Cheap to call
const created_at = ['avatars', 'posters', 'animations']
const has_history = ['statements', 'events']
export function is_history(itemid) {
  const parts = as_path_parts(itemid)
  if (has_history.includes(as_type(itemid)) && parts.length === 3) return true
  return false
}
export function as_filename(itemid) {
  let filename = itemid
  if (itemid.startsWith('/+')) filename = `/people${filename}`
  if (created_at.includes(as_type(itemid)) || is_history(itemid)) return `${filename}.html`
  else return `${filename}/index.html`
}
export function as_storage_path(itemid) {
  const path_parts = as_path_parts(itemid)
  let path = `/${path_parts[0]}`
  if (itemid.startsWith('/+')) path = `/people${path}`
  switch (path_parts.length) {
    case 0:
      return null
    case 1:
      return path
    default:
      return `${path}/${path_parts[1]}`
  }
}
export function as_directory_id(itemid) {
  const parts = as_path_parts(itemid)
  return `/${parts[0]}/${parts[1]}/`
}
export function as_path_parts(itemid) {
  const path = itemid.split('/')
  if (path[0].length === 0) path.shift()
  return path
}
export function as_author(itemid) {
  const path = as_path_parts(itemid)
  const author = path[0] || ''
  if (author.startsWith('+1')) return `/${path[0]}`
  else return null
}
export function as_type(itemid) {
  const path = as_path_parts(itemid)
  if (path[1]) return path[1]
  if (itemid.startsWith('/+')) return 'person'
  else return null
}
export function as_created_at(itemid) {
  const path = as_path_parts(itemid)
  return parseInt(path[2])
}
export function as_query_id(itemid) {
  return itemid.substring(2).replace('/', '-').replace('/', '-')
}
export function as_fragment(itemid) {
  return `#${as_query_id(itemid)}`
}
export function type_as_list(item) {
  // Returns a list even if loading the item fails
  // the microdata spec requires properties values to
  // single value and iterable
  if (!item) return []
  const list = item[as_type(item.id)]
  if (list && Array.isArray(list)) return list
  else if (list) return [list]
  else return []
}
