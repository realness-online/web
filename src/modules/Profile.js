import * as firebase from 'firebase/app'
import Item from '@/modules/Item'
import Storage from '@/modules/Storage'
import 'firebase/storage'
function get_url(mobile, type) {
  const path = `/people${mobile}/${type}.html`
  return firebase.storage().ref().child(path).getDownloadURL()
}
export default {
  load(phone_number) {
    // console.log(phone_number)
    return new Promise((resolve, reject) => {
      this.items(phone_number, 'person').then(items => {
        resolve(items[0])
      })
    })
  },
  items(mobile, type) {
    return new Promise((resolve, reject) => {
      get_url(mobile, type).then((url) => {
        fetch(url).then(response => {
          response.text().then((server_text) => {
            const server_as_fragment = Storage.hydrate(server_text)
            let items = Item.get_items(server_as_fragment)
            resolve(items)
          })
        })
      })
    })
  }
}
