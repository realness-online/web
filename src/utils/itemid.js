/** @typedef {import('@/types').Id} Id */
/** @typedef {import('@/types').Type} Type */
/** @typedef {import('@/types').Created} Created */
/** @typedef {import('@/types').Author} Author */
/** @typedef {import('@/types').Item} Item */
import { get, set } from 'idb-keyval'
import get_item from '@/utils/item'
import { DOES_NOT_EXIST } from '@/use/sync'
import { url } from '@/utils/serverless'
import { decompress_html } from '@/utils/upload-processor'
import { as_directory_id } from '@/persistance/Directory'
import { newest_timestamp_first } from '@/utils/sorting'
/**
 * @type {Type[]}
 */
const has_archive = ['posters']

/**
 * @type {Type[]}
 */
const has_history = ['statements', 'events']

// this leaves me, history, relations, and offline

/**
 * @param {Id} itemid
 * @returns {Promise<string>}
 */
export const as_filename = async itemid => {
  let filename = itemid
  if (itemid.startsWith('/+')) filename = `people${itemid}`

  if (has_archive.includes(as_type(itemid))) {
    const archive = await as_archive(itemid)
    console.log('archive name', archive)
    if (archive) return `${archive}.html.gz`
    return `${filename}.html.gz`
  } else if (is_history(itemid)) return `${filename}.html.gz`

  return `${filename}/index.html.gz`
}

/**
 * @param {Id} itemid
 * @returns {Promise<string | null>}
 */
export const load_from_network = async itemid => {
  const url = await as_download_url(itemid)

  if (url) {
    const response = await fetch(url)

    // Check Content-Encoding header
    const content_encoding = response.headers.get('Content-Encoding')
    const compressed_html = await response.arrayBuffer()
    let html = null
    // If no content encoding or 'identity', data is already decompressed
    if (!content_encoding || content_encoding === 'identity')
      html = new TextDecoder().decode(compressed_html)
    else html = await decompress_html(compressed_html)

    if (!html) return null
    console.log('storage', itemid)
    await set(itemid, html)
    return get_item(html)
  }
  return null
}

/**
 * @param {Id} itemid
 * @param {Author} me
 * @returns {Promise<Item | null>}
 */
export const load = async (itemid, me = localStorage.me) => {
  console.trace('load', itemid)
  let item
  if (~itemid.indexOf(me)) {
    item = localStorage.getItem(itemid)
    if (item) return get_item(item)
    else if (as_type(itemid) === 'relations') return []
  }
  const result = await get(itemid)
  item = get_item(result)
  if (item) return item
  try {
    item = await load_from_network(itemid, me)
  } catch (e) {
    if (e.code === 'storage/unauthorized') return null
    throw e
  }
  if (item) return item
  return null
}

/**
 * @param {Id} itemid
 * @param {Author} me
 * @returns {Promise<Type[]>}
 */
export const list = async (itemid, me = localStorage.me) => {
  try {
    const item = await load(itemid, me)
    if (item) return type_as_list(item)
    return []
  } catch {
    return []
  }
}

/**
 * @param {Id} itemid
 * @returns {Promise<string | null>}
 */
export const as_download_url = async itemid => {
  if (itemid.startsWith('/+/')) return null
  try {
    return await url(await as_filename(itemid))
  } catch (e) {
    if (e.code === 'storage/object-not-found') {
      console.trace(itemid, '=>', await as_filename(itemid))
      const index = (await get('sync:index')) || {}
      index[itemid] = DOES_NOT_EXIST
      await set('sync:index', index)
      return null
    }
    throw e
  }
}

/**
 * Splits an item ID path into its constituent parts
 * @param {Id} itemid - The full item path (e.g. '/+123456/statements/789')
 * @returns {[Author] | [Author, Type] | [Author, Type, Created]} Array of path parts where:
 *   [0] = author ID (e.g. '+123456')
 *   [1] = item type (e.g. 'statements', 'events', 'posters', etc.)
 *   [2] = created_at timestamp (if applicable)
 * @example
 * as_path_parts('/+123456/statements/789') // ['+123456', 'statements', '789']
 * as_path_parts('/+123456') // ['+123456']
 * as_path_parts('/+123456/statements/789/') // ['+123456', 'statements', null, '789']
 */
export const as_path_parts = itemid => {
  const path = itemid.split('/')
  if (path[0].length === 0) path.shift()
  if (itemid.endsWith('/')) {
    //is a directory
    const [author, type, created = null, archive = null] = path
    if (!created && !archive) return [author, type, 'index', null]
    if (created) return [author, type, null, created]
  }
  return path
}

/**
 * @param {Id} itemid
 * @returns {string | null}
 */
export const as_storage_path = itemid => {
  const path_parts = as_path_parts(itemid)
  let path = `/${path_parts[0]}`
  if (itemid.startsWith('/+')) path = `/people${path}`
  switch (path_parts.length) {
    case 0:
      return null
    case 1:
      return path
    default:
      return `${path}/${path_parts[1]}`
  }
}

/**
 * @param {Id} itemid
 * @returns {string | null}
 */
export const as_author = itemid => {
  const path = as_path_parts(itemid)
  const author = path[0] || ''
  if (author.startsWith('+1')) return `/${path[0]}`
  return null
}

/**
 * @param {Id} itemid
 * @returns {Type | null}
 */
export const as_type = itemid => {
  const path = as_path_parts(itemid)
  if (path[1]) return path[1]
  if (itemid.startsWith('/+')) return 'person'
  return null
}

/**
 * @param {Id} itemid
 * @returns {Created | null}
 */
export const as_created_at = itemid => {
  const path = as_path_parts(itemid)
  /** @type {1 | 2 | 3 | 4} */
  const path_length = path.length
  if (path_length === 3) return parseInt(path[2])
  if (path_length === 4) return parseInt(path[3])
  return null
}

/**
 * @param {Id} itemid
 * @returns {string}
 */
export const as_query_id = itemid =>
  itemid.substring(2).replace('/', '-').replace('/', '-')
//

/**
 * @param {Id} itemid
 * @returns {string}
 */
export const as_fragment_id = itemid => `#${as_query_id(itemid)}`

/**
 * @param {Item} item
 * @returns {Type[]}
 */
export const type_as_list = item => {
  // Returns a list even if loading the item fails
  // the microdata spec requires properties values to
  // single value and iterable
  if (!item) return []
  const list = item[as_type(item.id)]
  if (list && Array.isArray(list)) return list
  else if (list) return [list]
  return []
}
/**
 * @param {Id} itemid
 * @returns {boolean}
 */
export const is_history = itemid => {
  const parts = as_path_parts(itemid)
  if (has_history.includes(as_type(itemid)) && parts.length === 3) return true
  return false
}

/**
 * Determines if an item needs to be loaded from an archive
 * @param {Id} itemid - The ID of the item to check
 * @returns {Promise<string | null>} - Returns null if the item exists in the main items list,
 *                                    otherwise returns the archive path if found in an archive,
 *                                    or null if not found in either location
 */
export const as_archive = async itemid => {
  const directory = await get(as_directory_id(itemid))
  if (!directory) return null

  const { items = [], archive = [] } = directory
  const created = as_created_at(itemid)
  if (!created) return null

  if (items.includes(created.toString())) return null

  if (archive.includes(created))
    return `people${as_author(itemid)}/${as_type(itemid)}/${created}/${created}`

  const sorted_archive = [...archive].sort(newest_timestamp_first)

  let closest_timestamp = null // Find the closest archive timestamp that's GREATER than or EQUAL to the created timestamp
  for (const timestamp of sorted_archive)
    if (created <= timestamp) {
      closest_timestamp = timestamp
      break // Stop at the first timestamp that's greater than or equal to created
    }

  if (!closest_timestamp) return null
  return `people${as_author(itemid)}/${as_type(itemid)}/${closest_timestamp}/${created}`
}
