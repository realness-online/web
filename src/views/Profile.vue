<template>
  <section id="profile" class="page">
    <header>
      <realness-icon name="nothing" />
      <logo-as-link />
    </header>
    <div>
      <as-avatar v-if="person" :itemid="person.avatar" />
      <menu v-if="person">
        <as-download :itemid="person.avatar" />
        <as-messenger :itemid="person.id" />
      </menu>
    </div>
    <as-figure v-if="person" :person="person" :relations="relations" />
    <as-days v-slot="items" :posters="posters" :statements="statements">
      <template v-for="item in items">
        <poster-as-figure
          v-if="item.type === 'posters'"
          :key="slot_key(item.id)"
          :itemid="item.id" />
        <thought-as-article
          v-else
          :key="slot_key(item)"
          :statements="item"
          @show="thought_shown" />
      </template>
    </as-days>
  </section>
</template>
<script setup>
  import AsDays from '@/components/as-days'
  import LogoAsLink from '@/components/logo-as-link'
  import AsDownload from '@/components/download-vector'
  import AsFigure from '@/components/profile/as-figure'
  import AsAvatar from '@/components/posters/as-svg'
  import AsMessenger from '@/components/profile/as-messenger'
  import ThoughtAsArticle from '@/components/statements/as-article'
  import PosterAsFigure from '@/components/posters/as-figure'
  import RealnessIcon from '@/components/icon'
  import { from_e64 } from '@/use/profile'
  import { use as use_statements, slot_key } from '@/use/statements'
  import { use_posters } from '@/use/vector'
  import { use as use_person } from '@/use/people'
  import { ref, onMounted as mounted } from 'vue'
  import { useRoute as use_route } from 'vue-router'
  const route = use_route()
  const id = from_e64(route.params.phone_number)
  const {
    statements,
    thought_shown,
    for_person: statements_for_person
  } = use_statements()
  const { posters, for_person: posters_for_person } = use_posters()
  const { load_person, load_relations, person, relations } = use_person()
  mounted(async () => {
    await Promise.all([
      load_relations({ id: localStorage.me }),
      load_person(id),
      posters_for_person(id),
      statements_for_person(id)
    ])
    console.info('views:Profile', id)
  })
</script>
<style lang="stylus">
  section#profile
    padding: 0
    & > header
      height: 0
      padding: 0
      & > a
        -webkit-tap-highlight-color: blue
        z-index: 2
        top: inset(top)
        right: base-line
        position: absolute
    & > div
      position: relative
      overflow: hidden
      & > svg
        width: 100vw
        min-height: 100.5vh
      & > menu
        width: 100%
        display: flex
        justify-content: space-between
        z-index: 1
        position: absolute
        bottom: -(base-line * 2)
        bottom: 0
        padding: 0 base-line
        animation: absolute-slide-up
        animation-delay: 1.33s
        animation-duration: 0.35s
        animation-fill-mode: both
        & > a > svg
          fill: blue
    & > figure.profile
      padding: base-line
      & > svg
        border-radius: 0.66em
        width: base-line * 2
        height: base-line * 2
        border-radius: base-line
      & > figcaption > menu
        a.status
          position: absolute
          animation: absolute-slide-down
          animation-delay: 1.33s
          animation-duration: 0.35s
          animation-fill-mode: both
          svg.add
            width: base-line * 2
            height: base-line * 2
          &.relation
            svg.add
              width: 0
              height: 0
            svg.finished
              width: base-line * 2
              height: base-line * 2
        a.phone
          display: none
    & > section.as-days
      & > article.day
        @media (prefers-color-scheme: dark)
          & > header > h4, figure.poster > svg.background
            color: blue
        figure.poster
          & > figcaption > menu
            & > a.download svg
              fill: blue
            & > a.phone
              display: none
            & > a.profile
            & > address
              & > time
                font-size: max-font
                color: blue
              & > h3
                display: none
            & > svg
              display: none
</style>
