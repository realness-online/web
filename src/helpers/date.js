import { as_created_at } from '@/helpers/itemid'
const this_year = Date.parse(new Date().getFullYear())
const format_as_time = {
  hour: 'numeric',
  minute: 'numeric',
  hour12: true
}
const format_as_day = {
  weekday: 'long',
  day: 'numeric',
  month: 'long'
}
const format_as_day_and_year = {
  weekday: 'long',
  day: 'numeric',
  month: 'long',
  year: 'numeric'
}
const format_as_day_and_time = {
  weekday: 'long',
  day: 'numeric',
  month: 'long',
  hour: 'numeric',
  minute: 'numeric',
  hour12: true
}
export default {
  as_time (date, format = format_as_time) {
    const time = new Date(date)
    return time.toLocaleString('en-US', format) // TODO: get country code from browser
  },
  as_day (date) {
    let day
    if (Date.parse(date) < this_year) {
      day = this.as_time(date, format_as_day_and_year)
    } else day = this.as_time(date, format_as_day)
    if (this.is_same_day(date, new Date())) day = 'Today'
    return day
  },
  as_day_and_time (date) {
    return this.as_time(date, format_as_day_and_time)
  },
  is_same_day (d1, d2) {
    d1 = new Date(d1)
    d2 = new Date(d2)
    return d1.getFullYear() === d2.getFullYear() &&
      d1.getDate() === d2.getDate() && d1.getMonth() === d2.getMonth()
  },
  day_name (date) {
    date = new Date(date)
    return `${date.getMonth() + 1}/${date.getDate()}/${date.getFullYear()}`
  },
  id_as_day (itemid) {
    return new Date(as_created_at(itemid))
  }

}
