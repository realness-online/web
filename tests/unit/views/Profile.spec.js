import { shallowMount } from '@vue/test-utils'
import firebase from 'firebase/app'
import 'firebase/auth'
import Profile from '@/views/Profile'
import flushPromises from 'flush-promises'
import * as itemid from '@/helpers/itemid'
const person = require('fs').readFileSync('./tests/unit/html/person.html', 'utf8')
const fetch = require('jest-fetch-mock')
const user = { phoneNumber: '/+16282281824' }

describe('@/views/Profile.vue', () => {
  describe('Renders', () => {
    it('Shows profile information for a phone number', async () => {
      firebase.user = user
      fetch.resetMocks()
      fetch.mockResponseOnce(person)
      jest.spyOn(itemid, 'as_directory').mockImplementationOnce(() => {
        return { items: ['559666932867'] }
      })
      const $route = { params: { phone_number: '+14151231234' } }
      const wrapper = shallowMount(Profile, { global: { mocks: { $route } } })
      await flushPromises()
      expect(wrapper.element).toMatchSnapshot()
      firebase.user = null
    })
    it('Handles there not being posters', async () => {
      firebase.user = user
      fetch.resetMocks()
      fetch.mockResponseOnce(person)
      jest.spyOn(itemid, 'as_directory').mockImplementationOnce(() => {
        return { items: [] }
      })
      const $route = { params: { phone_number: '+14151231234' } }
      const wrapper = shallowMount(Profile, { global: { mocks: { $route } } })
      await flushPromises()
      expect(wrapper.element).toMatchSnapshot()
      firebase.user = null
    })
  })
})
