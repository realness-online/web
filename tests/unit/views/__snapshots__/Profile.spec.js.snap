// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@/views/Profile.vue Shows profile information for a phone number 1`] = `
<section
  class="page"
  id="profile"
>
  <header>
    <svg />
     
    <!---->
  </header>
   
  <!---->
   
  <!---->
   
  <!---->
</section>
`;

exports[`@/views/Profile.vue shows profile information for a phone number 1`] = `
<section
  class="page"
  id="profile"
>
  <header>
    <svg />
     
    <!---->
  </header>
   
  <!---->
   
  <!---->
   
  <!---->
</section>
`;
